pkg/machines/components/vcpuModal.jsx

```diff
@@ -0,0 +1,192 @@
+import React from 'react';
+import cockpit from 'cockpit';
+
+import { show_modal_dialog } from 'cockpit-components-dialog.jsx';
+import SelectComponent from 'cockpit-components-select.jsx';
+import InfoRecord from './infoRecord.jsx';
+import { setVCPUSettings } from "../actions.es6";
+
+React;
+
+const _ = cockpit.gettext;
+
+const dividers = (num) => {
+    const divs = [1];
+
+    for (let i = 2; i < num; i++) {
+        if (num % i === 0) {
+            divs.push(i);
+        }
+    }
+
+    if (num > 1) {
+        divs.push(num);
+    }
+
+    return divs;
+}
+
+const Select = function ({ id, items, onChange, value }) {
+    return ( <SelectComponent.Select id={id} initial={value} onChange={onChange}>
+        {items.map((t) => (
+            <SelectComponent.SelectEntry data={t}>{t}</SelectComponent.SelectEntry>
+        ))}
+    </SelectComponent.Select>);
+}
+
+class VCPUModalBody extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            max: props.vcpus.max || 1,
+            count: parseInt(props.vcpus.count) || 1,
+            sockets: props.cpu.topology.sockets || 1,
+            threads: props.cpu.topology.threads || 1,
+            cores: props.cpu.topology.cores || 1
+        };
+        this.onMaxChange = this.onMaxChange.bind(this);
+        this.onCountSelect = this.onCountSelect.bind(this);
+        this.onSocketChange = this.onSocketChange.bind(this);
+        this.onThreadsChange = this.onThreadsChange.bind(this);
+        this.onCoresChange = this.onCoresChange.bind(this);
+        props.onChange(this.state);
```

Comment:
I see why you mix props and state here, as in this case props just define the initial state. But this feels wrong to me - once the user changes the values manually, there should not be any property updates, and even if they were they should not override the user changes.

Reply:
They not overwrite user changes, it only update local variable, which is used for saving new VCPU settings. You can see it on line 184.

Reply:
I'm afraid I don't understand.. `this.state` is *the* state of the React component, and one can't just change it like that - this should call `setState()`. But moreover, state is modified by the user in the dialog, not by property changes from outside. What do you mean with "local variable"?

Reply:
I mean, that if user change some value, then there is no way to change it from outside. `onChange` doesn't change state of React Component, thus it doesn't change manually changed values.